base/multidimensional.jl

```diff
@@ -956,12 +956,22 @@ julia> y
 copyto!(dest, src)
 
 function copyto!(dest::AbstractArray{T1,N}, src::AbstractArray{T2,N}) where {T1,T2,N}
-    checkbounds(dest, axes(src)...)
     src′ = unalias(dest, src)
-    for I in eachindex(IndexStyle(src′,dest), src′)
-        @inbounds dest[I] = src′[I]
+    # fastpath for equal axes (#34025)
+    if axes(dest) == axes(src)
+        for I in eachindex(IndexStyle(src′,dest), src′)
+            @inbounds dest[I] = src′[I]
+        end
+    # otherwise enforce linear indexing
+    else
+        linds = eachindex(IndexLinear(), src)
+        checkbounds(dest, first(linds))
+        checkbounds(dest, last(linds))
+        for I in linds
```

Comment:
The docs talk about "the first `n` elements" so I think this needs to ensure that we target the beginning of both arrays. Here's an example where the above fails:

```julia
a = fill(1, 5)
r = Base.IdentityUnitRange(-1:1)
copyto!(a, r)
```